File: 82.RemoveDupSortedListII/RemDupListII.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        current = head
        
        while current:
            # Check if current node is a duplicate
            if current.next and current.val == current.next.val:
                # Skip all duplicates
                while current.next and current.val == current.next.val:
                    current = current.next
                # Link previous node to the next non-duplicate node
                prev.next = current.next
            else:
                # Move prev pointer only if no duplicates were found
                prev = prev.next
            
            # Move to the next node
            current = current.next
        
        return dummy.next

File: 82.RemoveDupSortedListII/RemDupListII.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function deleteDuplicates(head: ListNode | null): ListNode | null {
    if (!head) return null;

    const dummy = new ListNode(0);
    dummy.next = head;
    let prev = dummy;
    let current = head;

    while (current) {
        let isDuplicate = false;

        // Check for duplicates
        while (current.next && current.val === current.next.val) {
            isDuplicate = true;
            current = current.next;
        }

        if (isDuplicate) {
            // Skip all duplicates
            prev.next = current.next;
        } else {
            // Move prev pointer only if no duplicates
            prev = prev.next;
        }

        current = current.next;
    }

    return dummy.next;
}
